### ISSUE TITLE
eslint no-misused-promise in notFoundHandler

### ISSUE THREAD

**Issue by mindrunner** [fastify/fastify-rate-limit#355]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.25.2

### Plugin version

9.0.1

### Node.js version

20.10.0

### Operating system

Linux

### Operating system version (i.e. 20.04, 11.3, 10)

-

### Description

I am rate limiting 404 errors as documented here:
https://github.com/fastify/fastify-rate-limit#preventing-guessing-of-urls-through-404s

`notFoundHandler` expects a `preValidationHookHandler`, but `fastify.rateLimit()` is a `preHandlerAsyncHookHandler`

### Steps to Reproduce

Use
```
fastify.setNotFoundHandler({
  preHandler: fastify.rateLimit()
}, function (request, reply) {
  reply.code(404).send({ hello: 'world' })
})
```

- Enable `no-misused-promise` eslint rule

### Expected Behavior

No erorrs / warnings thrown

**Comment by mindrunner** [comment]
The problem here is not the handler itself, but the `preHandler` which does not match:

![image](https://github.com/fastify/fastify-rate-limit/assets/1413542/5df9473e-f65c-4baa-bb12-80d4c1aee9ec)

**Comment by mcollina** [comment]
Ah it might be a bug in our types. That does not return a promise. That returns an async function.

**Comment by mindrunner** [comment]
Yea that was my assumption. Can have a look into this later.

**Comment by mindrunner** [comment]
I am not sure how to fix this. `preHandler` expexts a `preHandlerHookHandler`, but `rateLimit` returns a `preHandlerAsyncHookHandler`

**Comment by mcollina** [comment]
Can you share a reproduction for this? So somebody can quickly work for a resolution.

**Comment by mindrunner** [comment]
See 'Steps to reproduce' in Original Post. What else would you need to reproduce?

**Comment by mcollina** [comment]
We often need a [reproducible example](https://stackoverflow.com/help/minimal-reproducible-example), e.g. some code that allows someone else to recreate your problem by just copying and pasting it. If it involves more than a couple of different file, create a new repository on GitHub and add a link to that.

TL;DR Set up a tiny repository that shows the problem.

**Comment by mindrunner** [comment]
app.ts:
```
import Fastify from "fastify";

const main = async () => {
  const fastify = Fastify();
  await fastify.register(import("@fastify/rate-limit"), {
    max: 100,
    timeWindow: "1 minute",
  });

  fastify.setNotFoundHandler(
    {
      preHandler: fastify.rateLimit(),
    },
    function (request, reply) {
      reply.code(404).send({ hello: "world" });
    },
  );
};

main().then(() => {});
```

tsconfig.json
```
{
  "name": "rl-fastify",
  "version": "1.0.0",
  "description": "",
  "dependencies": {
    "fastify": "^4.25.2",
    "@fastify/rate-limit": "^9.1.0"
  },
  "devDependencies": {
    "@typescript-eslint/eslint-plugin": "^6.16.0",
    "@typescript-eslint/parser": "^6.16.0",
    "eslint": "^8.56.0"
  }
}
```

.eslintrc.js
```
{
  "compilerOptions": {
    "target": "es2016",
    "module": "commonjs",
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "skipLibCheck": true
  },
  "include": [
    "app.ts",
    ".eslintrc.js"
  ]
}
```


Steps to reproduce:

- copy all `3` files to same directory
- run `pnpm install`
- run `eslint .`

Actual output:

```
  12:19  error  Promise-returning function provided to property where a void return was expected  @typescript-eslint/no-misused-promises

✖ 1 problem (1 error, 0 warnings)
```

Expected output:

```
<empty> (e.g. eslint does not show any errors)
```

**Comment by mcollina** [comment]
I think you got your file names wrong. I recommend you to upload everything to a repository.

**Comment by mindrunner** [comment]
Sorry, accidentally hit save before I was ready.

**Comment by mindrunner** [comment]
https://github.com/mindrunner/rl-fastify

**Comment by mindrunner** [comment]
updated repo. pls try again

**Comment by mcollina** [comment]
Thanks, that was helpful.

Here is the fix: https://github.com/fastify/fastify/pull/5258
